hash.c

```diff
@@ -6528,68 +6539,235 @@ env_update(VALUE env, VALUE hash)
+ *  - A \Hash key can be (almost) any object.
  *
- *  Hashes enumerate their values in the order that the corresponding keys
- *  were inserted.
+ *  === \Hash \Data Syntax
  *
- *  A Hash can be easily created by using its implicit form:
+ *  The older syntax for \Hash data uses the "hash rocket," <tt>=></tt>
+ *  (known in some languages as the <i>fat comma.</i>):
  *
- *    grades = { "Jane Doe" => 10, "Jim Doe" => 6 }
+ *    h = {:foo => 0, :bar => 1, :baz => 2}
+ *    h # => {:foo=>0, :bar=>1, :baz=>2}
  *
- *  Hashes allow an alternate syntax for keys that are symbols.
- *  Instead of
  *
- *    options = { :font_size => 10, :font_family => "Arial" }
  *
- *  You could write it as:
+ *  Alternatively, but only for a \Hash key that's a \Symbol,
+ *  you can use a newer JSON-style syntax,
+ *  where each bareword becomes a \Symbol:
  *
- *    options = { font_size: 10, font_family: "Arial" }
+ *    h = {foo: 0, bar: 1, baz: 2}
+ *    h # => {:foo=>0, :bar=>1, :baz=>2}
  *
- *  Each named key is a symbol you can access in hash:
+ *  You can also use a \String in place of a bareword:
  *
- *    options[:font_size]  # => 10
+ *    h = {'foo': 0, 'bar': 1, 'baz': 2}
+ *    h # => {:foo=>0, :bar=>1, :baz=>2}
  *
- *  A Hash can also be created through its ::new method:
+ *  And you can mix the styles:
  *
- *    grades = Hash.new
- *    grades["Dorothy Doe"] = 9
+ *    h = {foo: 0, :bar => 1, 'baz': 2}
+ *    h # => {:foo=>0, :bar=>1, :baz=>2}
  *
- *  Accessing a value in a Hash requires using its key:
+ *  But it's an error to try the JSON-style syntax
+ *  for a key that's not a bareword or a String:
  *
- *    puts grades["Jane Doe"] # => 0
+ *    # Raises SyntaxError (syntax error, unexpected ':', expecting =>)
+ *    h = {0: 'zero'}
  *
  *  === Common Uses
  *
- *  Hashes are an easy way to represent data structures, such as
+ *  You can use a \Hash to give names to objects:
+ *
+ *    matz = {name: 'Matz', language: 'Ruby'}
+ *    matz # => {:name=>"Matz", :language=>"Ruby"}
+ *
+ *  You can use a \Hash to give names to method arguments:
  *
- *    books         = {}
- *    books[:matz]  = "The Ruby Programming Language"
- *    books[:black] = "The Well-Grounded Rubyist"
+ *    def some_method(hash)
+ *      p hash
+ *    end
+ *    some_method({foo: 0, bar: 1, baz: 2}) # => {:foo=>0, :bar=>1, :baz=>2}
+ *
+ *  Note: when the last argument in a method call is a \Hash,
+ *  the curly braces may be omitted:
  *
- *  Hashes are also commonly used as a way to have named parameters in
- *  functions. Note that no brackets are used below. If a hash is the last
- *  argument on a method call, no braces are needed, thus creating a really
- *  clean interface:
+ *    some_method(foo: 0, bar: 1, baz: 2) # => {:foo=>0, :bar=>1, :baz=>2}
  *
- *    Person.create(name: "John Doe", age: 27)
+ *  You can use a Hash to initialize an object:
  *
- *    def self.create(params)
- *      @name = params[:name]
- *      @age  = params[:age]
+ *    class Dev
```

Comment:
I personally don't think we should be encouraging this style.

Reply:
Fixed.